assembly/src/ast/module.rs

```diff
@@ -1,321 +1,718 @@
+    /// Get the fully-qualified name of this module, e.g. `std::math::u64`
+    pub fn path(&self) -> &LibraryPath {
+        &self.path
+    }
 
-    /// Clears the source locations from this module.
-    pub fn clear_locations(&mut self) {
-        self.local_procs.iter_mut().for_each(|p| p.clear_locations())
+    /// Get the namespace of this module, e.g. `std` in `std::math::u64`
+    pub fn namespace(&self) -> &LibraryNamespace {
+        self.path.namespace()
     }
 
-    // SERIALIZATION / DESERIALIZATION
-    // --------------------------------------------------------------------------------------------
+    /// Returns true if this module belongs to the provided namespace.
+    pub fn is_in_namespace(&self, namespace: &LibraryNamespace) -> bool {
+        self.path.namespace() == namespace
+    }
+
+    /// Get the module documentation for this module, if it was present in the source code the
+    /// module was parsed from
+    pub fn docs(&self) -> Option<Span<&str>> {
+        self.docs.as_ref().map(|spanned| spanned.as_deref())
+    }
 
-    /// Returns byte representation of this [ModuleAst].
+    /// Get the type of module this represents:
     ///
-    /// The serde options are NOT serialized - the caller must keep track of the serialization
-    /// options used.
-    pub fn write_into<R: ByteWriter>(&self, target: &mut R, options: AstSerdeOptions) {
-        // asserts below are OK because we enforce limits on the number of procedure and length of
-        // module docs in the module parser
-
-        // serialize docs
-        match &self.docs {
-            Some(docs) => {
-                assert!(docs.len() <= u16::MAX as usize, "docs too long");
-                target.write_u16(docs.len() as u16);
-                target.write_bytes(docs.as_bytes());
-            }
-            None => {
-                target.write_u16(0);
+    /// See [ModuleKind] for details on the different types of modules.
+    pub fn kind(&self) -> ModuleKind {
+        self.kind
+    }
+
+    /// Returns true if this module is an executable module.
+    #[inline(always)]
+    pub fn is_executable(&self) -> bool {
+        self.kind.is_executable()
+    }
+
+    /// Returns true if this module is a kernel module.
+    #[inline(always)]
+    pub fn is_kernel(&self) -> bool {
+        self.kind.is_kernel()
+    }
+
+    /// Returns true if this module has an entrypoint procedure defined,
+    /// i.e. a `begin`..`end` block.
+    pub fn has_entrypoint(&self) -> bool {
+        self.index_of(|p| p.is_main()).is_some()
+    }
+
+    /// Get an iterator over the procedures defined in this module.
+    ///
+    /// The entity returned is an [Export], which abstracts over locally-defined procedures and
+    /// re-exported procedures from imported modules.
+    pub fn procedures(&self) -> core::slice::Iter<'_, Export> {
+        self.procedures.iter()
+    }
```

Comment:
Question: is there a benefit to specifying the concrete iterator type instead of doing something like `impl Iterator`?

Reply:
In general, you have a choice between exposing the concrete type, expressing the full set of traits implemented by the concrete iterator generically, or constraining the functionality of the iterator to a subset of those traits.

For example, `slice::Iter` implements `ExactSizeIterator`, `DoubleEndedIterator`, `FusedIterator`, and `TrustedLen`. Additionally, the `Iter` type itself exposes an `as_slice` method which is useful for accessing the rest of the contents of the iterator as a slice. To expose an iterator with equivalent functionality while hiding the concrete implementation, we'd need the following:

```rust
pub fn procedures(&self) -> impl DoubleEndedIterator<Item = &Export> + ExactSizeIterator + FusedIterator + TrustedLen + AsRef<[Export]> + '_ {
    ...
}
```

In practice, that's too much work, so we might just expose it as just:

```rust
pub fn procedures(&self) -> impl DoubleEndedIterator<Item = &Export> + '_ {
   ...
}
```

This gets us the reversible iterator, but we lose the the benefits provided by the other traits.

In general, I think it's preferable to use RPIT types (return-position impl trait) rather than the concrete type, but with some iterators, particularly those for slices/vecs/str, the concrete implementations provide a lot of specialization and rich functionality, that is difficult to match with RPIT types, and in those cases I tend to prefer exposing the concrete iterator. It does mean that the concrete type becomes part of your public API, but any significant refactoring to use a different underlying data structure here (such as an intrusive collection) would likely result in breaking changes to the API anyway.

But it's always a judgement call, in this case, it probably isn't critical that we expose the concrete type, based on how it is currently being used.